notif [nfc]: Update docs for testing E2EE notifications (#2304)
zulip/zulip-flutter: docs/howto/push-notifications.md

   You can now delete the other files `/tmp/apns.key`, `/tmp/apns.csr`,
   and `/tmp/aps_development.cer`.

4. Generate a bouncer public key and the corresponding private key
   by running the following management command on the dev server:
   ```
   $ ./manage.py manage_push_registration_encryption_keys --add

>    ```

Reviewer: Ah, well, Claude had a good idea here. 🙂 This is from output of the `/code-review` slash-command:

```json
    {
      "file": "docs/howto/push-notifications.md",
      "line": 202,
      "summary": "The doc never shows the step-4 command's output, how to display the public-key value, which of the two generated
  values to copy, or the required format.",
      "failure_scenario": "The constant is base64Decode('mm4F...REM=') at push_device.dart:282; step 4 generates a public AND private
  key with no shown output. A reader reaches step 6 needing 'the public key we generated' with no command to print it and no format
  guidance — stuck mid-procedure, or pasting the wrong value/encoding into base64Decode, yielding a FormatException or undecryptable
  registrations."
    },
```

(I think this tool was designed to push comments to GitHub PRs, and I think that explains the JSON formatting.)

I think I'd still prefer not to paste the keys themselves (they look like secrets, which seems confusing). But let's add back a bit, here's an attempt:

```diff
--- a/docs/howto/push-notifications.md
+++ b/docs/howto/push-notifications.md
@@ -195,13 +195,23 @@ on that is
    $ ./manage.py manage_push_registration_encryption_keys --add
    ```
 
+   Output is a public/private keypair, base46-encoded:
+
+   ```
+   Added a new key pair:
+   - Public key: <etc.>
+   - Private key: <etc.>
+   ```
+
+   You'll use the "Public key" value in a later step.
+
 5. Restart `tools/run-dev` to let the server pick up the changes.
    It should automatically see the file `zproject/apns-dev.pem`
    and use it to communicate with the APNs sandbox server.
 
-6. Then in the app, replace the hardcoded `bouncerPublicKey` in
-   `lib/model/push_device.dart` with the public key we generated
-   in step 4.
+6. Then in the app, source `bouncerPublicKey` in
+   `lib/model/push_device.dart` from the "Public key" value from step 4,
+   instead of the corresponding value for the production bouncer.
 
    Then hot restart the app to let the device registration take place
    with the new public key.
```

5. Restart `tools/run-dev` to let the server pick up the changes.
   It should automatically see the file `zproject/apns-dev.pem`

>    and use it to communicate with the APNs sandbox server.

Reviewer: ```suggestion
5. Restart `tools/run-dev` to let the server pick up the changes:
   the push registration encryption keys generated in step 4, and
   the file `zproject/apns-dev.pem` for communicating with the
   APNs sandbox server.
```

6. Then in the app, replace the hardcoded `bouncerPublicKey` in
   `lib/model/push_device.dart` with the public key we generated
   in step 4.

   Then hot restart the app to let the device registration take place
   with the new public key.

> You should now be getting notifications on your iOS development build!

Reviewer: ```suggestion
You should now be getting notifications on your iOS development build!
(On a real device, though, a debug build shows E2EE notifications
only as "New notification"; see the troubleshooting entry below.)
```

And then maybe make the `"New notification" notification when running debug app on real device` section come first in the list of troubleshooting entries, because I feel like it's likely to bite and is also pretty easy to work around.

   You can now delete the other files `/tmp/apns.key`, `/tmp/apns.csr`,
   and `/tmp/aps_development.cer`.

4. Generate a bouncer public key and the corresponding private key
   by running the following management command on the dev server:
   ```
   $ ./manage.py manage_push_registration_encryption_keys --add
   Added a new key pair:
   - Public key:  5gByix53AzwEnMTN4H/nbaEehJMMkY3tHNYKBlMSxj8=
   - Private key: iYnEVKNajGjCEodfARRJu+QUDisTMgZQ4+ADy4dkBWY=
   ```

5. Restart `tools/run-dev` to let the server pick up the changes.
   It should automatically see the file `zproject/apns-dev.pem`
   and use it to communicate with the APNs sandbox server.

6. Then in the app, replace the hardcoded `bouncerPublicKey` in
   `lib/model/push_device.dart` with the public key we generated

>    in the step 4.

Reviewer: nit: "in step 4"

   You can now delete the other files `/tmp/apns.key`, `/tmp/apns.csr`,
   and `/tmp/aps_development.cer`.

4. Generate a bouncer public key and the corresponding private key
   by running the following management command on the dev server:
   ```
   $ ./manage.py manage_push_registration_encryption_keys --add
   Added a new key pair:
   - Public key:  5gByix53AzwEnMTN4H/nbaEehJMMkY3tHNYKBlMSxj8=
   - Private key: iYnEVKNajGjCEodfARRJu+QUDisTMgZQ4+ADy4dkBWY=

>    ```

Reviewer: If we omit the example output and have step 6 just say "replace […] with the public key from step 4", then I think it's still clear enough, and readers don't have to think through whether the specific strings "5gByix53AzwEnMTN4H/nbaEehJMMkY3tHNYKBlMSxj8=" and "iYnEVKNajGjCEodfARRJu+QUDisTMgZQ4+ADy4dkBWY=" are important.

Reviewer: Bump on omitting the example output here; my suggestion is to just list the command: `./manage.py manage_push_registration_encryption_keys --add`.

   You can now delete the other files `/tmp/apns.key`, `/tmp/apns.csr`,
   and `/tmp/aps_development.cer`.

4. Generate a bouncer public key and the corresponding private key
   by running the following management command on the dev server:
   ```
   $ ./manage.py manage_push_registration_encryption_keys --add
   ```

>    Output is a public/private keypair, base46-encoded:

Reviewer: nit: "base64-encoded"

chat.zulip.org, so we can debug.

> ##### Logs for NotificationService don't show up in `flutter run/logs`

Reviewer: nit: nonexistent command `flutter run/logs`

```suggestion
##### Logs for NotificationService don't show up in `flutter run`/`flutter logs`
```

Logs will not show up in `flutter run` or `flutter logs` (probably)
because the code running in the NotificationService target runs in
a separate process. And Flutter's tooling is not routing/reading

> logs from that process.

Reviewer: Can we cut the "(probably)", or is there still something to pin down here?

chat.zulip.org, so we can debug.

##### Logs for NotificationService don't show up in `flutter run/logs`

Logs will not show up in `flutter run` or `flutter logs` (probably)
because the code running in the NotificationService target runs in
a separate process. And Flutter's tooling is not routing/reading
logs from that process.

One workaround is to use Console.app for reading the logstream from
the NotificationService process by following these steps:

> 1. Open the Console app

Reviewer: nit: final punctuation, like in the other steps

Reviewer: nit: Also let's be consistent between "Console.app" and "the Console app".

4. Press Start (with a "play" icon) to start streaming the logs
   with the specified filter.
5. Receive a notification, and that will result in some logs being
   emitted.

##### "New notification" notification when running debug app on real device

When running the debug build (`flutter run`) of the app on a
real iPhone, the FlutterEngine in the NotificationService extension
will not start, which will result in the NotificationService
extension falling back to displaying the original notification's
"alert", which for E2EE notifications says "New notification".
And the following error will be logged:
```
Cannot create a FlutterEngine instance in debug mode without Flutter tooling or Xcode.

To launch in debug mode in iOS 14+, run flutter run from Flutter tools, run from an IDE with a Flutter IDE plugin or run the iOS project from Xcode.
Alternatively profile and release mode apps can be launched from the home screen.
```

This is probably because on real iPhone, iOS only allows JIT-ed
code to run if it's being pushed via the attached debugger.
Which is generally the case with the Runner process (the main app)
when doing a `flutter run`. But not the case with secondary

> FlutterEngine running in the NotificationService extension.

Reviewer: Reading this paragraph again, I want to check if I'm parsing it correctly.

- What does "probably" mean?
  - Are you offering a likely theory based on facts you're certain about (i.e. you know the OS's behavior from careful observation or by reading docs, e.g. on Apple's security model),
  - or are you also unsure about the facts?

Either is fine, but let's make the answer clear in the text, and perhaps downgrade "probably" to something weaker if you're not sure how well the conclusion is supported.

Author: It was mostly based on observations from a previous upstream bug, specifically: https://github.com/flutter/flutter/issues/163984

Anyway, changed the language a bit.

To work around this, run the release version of the app
(`flutter run --release`), or run the debug version on the
iOS Simulator.

##### Can't hot reload/restart the Dart code running in NotificationService

The Flutter tooling doesn't attach to the secondary FlutterEngine
running in the NotificationService extension when doing a

> `flutter run`. There are no know workarounds as of now for this,

Reviewer: nit: "known"

only way to update the running code is by doing a complete build

> again (stopping `flutter run` starting it again).

Reviewer: also nit: comma splice and missing "the"; how about:

> There are no known workarounds for this; the only way to update the running code is […].

Reviewer: nit: "stopping `flutter run` _and_ starting it again"

chat.zulip.org, so we can debug.

> ##### Logs for NotificationService doesn't show up in `flutter run/logs`

Reviewer: nit: s/doesn't/don't/

Logs will not show up in `flutter run` or `flutter logs` (probably)

> because the code running in NotificationService target runs in a

Reviewer: nit: "in the NotificationService target"

(similarly "the NotificationService extension"; "the NotificationService process", etc., in a few other places)

separate process. And Flutter's tooling is not routing/reading logs
from that process.

One workaround is to use Console.app for reading the logstream from

> `NotificationService` process by following these steps:

Reviewer: nit: let's consistently write either NotificationService or `NotificationService` in this new material.

1. Open the Console app
2. From the sidebar select the device or the simulator that you

>    desire to read the logs from.

Reviewer: nit: verbose; how about

```md
2. Select your device or simulator in the left sidebar.
```

3. In the search bar type "NotificationService" and press enter.
   That will create an entry like "Any: NotificationService" in the
   search bar.
4. Right click on that entry, and select "Process" instead of "Any",

>    to filter logs by process name.

Reviewer: I think steps 3 and 4 can be simplified into one step, like

```md
3. In the search bar, type "process:NotificationService" and press Return/Enter.
```

5. Press the play button to start streaming the logs with the

>    specified filter.

Reviewer: nit:

```md
4. Press Start (with a "play" icon) […].
```

6. Receive a notification, and that will result in some logs being

>    emitted in the Console app.

Reviewer: nit: no need to mention the Console app again; e.g.

```md
6. Receive a notification and watch for logs from NotificationService.
```

Could also add a note that NotificationService only processes notifications that will cause an alert; this comes from the yellow banner in https://developer.apple.com/documentation/usernotifications/modifying-content-in-newly-delivered-notifications .

##### "New notification" notification when running debug app on real device

When running the debug build (`flutter run`) of the app on a real
iPhone, the FlutterEngine in NotificationService extension will not
start. Resulting in falling back to displaying the original
notifications "alert", which for E2EE notifications say
"New notification". And the following error will be logged in

> the Console app:

Reviewer: nits:

- "Resulting in […]" is a sentence fragment; either make a complete sentence (with subject and verb) or combine with previous sentence.
- 'the original notification's "alert"', right (note the apostrophe)

```
Cannot create a FlutterEngine instance in debug mode without Flutter tooling or Xcode.

To launch in debug mode in iOS 14+, run flutter run from Flutter tools, run from an IDE with a Flutter IDE plugin or run the iOS project from Xcode.
Alternatively profile and release mode apps can be launched from the home screen.
```

This is probably because on real iPhone, iOS only allows JIT-ed
code to run if it's being pushed via the attached debugger.
Which is generally the case with the Runner process (the main app)
when doing a `flutter run`. But not the case with secondary
FlutterEngine running in the NotificationService extension.

> To workaround this, run the release version of the app

Reviewer: nit: "To work around" (the single word "workaround" is a noun, meaning the thing you do to work around a problem)

code to run when a debugger is attached to the process.
This is generally the case for the Runner process (the main app)
when doing `flutter run`.
The NotificationService extension runs in a separate process,
so its debug-mode FlutterEngine will not start without
a debugger attached.

To work around this, run the release version of the app
(`flutter run --release`), or run the debug version on the
iOS Simulator.

##### Logs for NotificationService don't show up in `flutter run`/`flutter logs`

Logs will not show up in `flutter run` or `flutter logs`
because the code running in the NotificationService target runs in
a separate process. And Flutter's tooling is not routing/reading
logs from that process.

One workaround is to use the Console app for reading the logstream
from the NotificationService process by following these steps:
1. Open the Console app.
2. Select your device or simulator in the left sidebar.
3. In the search bar, type "process:NotificationService" and press
   Return/Enter.
4. Press Start (with a "play" icon) to start streaming the logs
   with the specified filter.
5. Receive a notification, and that will result in some logs being

>    emitted.

Reviewer: There are some `self.logger.debug` calls in NotificationService.swift. Does that mean the user should also enable "Action > Include Debug Messages"? (Should they also enable "Action > Include Info Messages"?) Untested, but seems reasonable; Claude flagged this.

Author: No, those logs were showing up in Console app without changing any other settings.
